bba/bba.go

```diff
@@ -211,19 +219,29 @@ func (bba *BBA) tryoutAgreement() {
 	}
 	if len(binList) > 1 {
 		bba.Log("bin value set size is larger than one")
-		bba.est.Set(cleisthenes.Binary(coin))
-		bba.advanceRoundChan <- struct{}{}
+		bba.agreementFailed(cleisthenes.Binary(coin))
 		return
 	}
-
 	if binList[0] != cleisthenes.Binary(coin) {
 		bba.Log("bin value set value is different with coin value")
-		bba.est.Set(binList[0])
-		bba.advanceRoundChan <- struct{}{}
+		bba.agreementFailed(binList[0])
 		return
 	}
-	bba.dec.Set(binList[0])
+	bba.agreementSuccess(binList[0])
+}
+
+func (bba *BBA) agreementFailed(estValue cleisthenes.Binary) {
+	bba.est.Set(estValue)
+	bba.advanceRoundChan <- struct{}{}
+}
+
+func (bba *BBA) agreementSuccess(decValue cleisthenes.Binary) {
+	bba.dec.Set(decValue)
 	bba.done.Set(true)
+	bba.binInputChan.Send(cleisthenes.BinaryMessage{
+		Member: bba.owner,
+		Binary: bba.dec.Value(),
```

Comment:
is it tread-safe?

Reply:
yes because it is channel there's no need lock/unlock

also for safety I tested with sample code like this

```
package main

import (
	"fmt"
	"sync"
)

func main() {
	wg := sync.WaitGroup{}
	buffer := make(chan struct{})
	wg.Add(100)
	go func() {
		i := 0
		for {
			select {
			case <-buffer:
				fmt.Println("received", i)
				i++
				wg.Done()
				if i == 100 {
					return
				}
			}
		}

	}()

	for i := 0; i < 10; i++ {
		go func() {
			for i := 0; i < 10; i++ {
				buffer<- struct{}{}
			}
		}()
	}
	wg.Wait()
}
```

Reply:
Oh, i was mistaken just for `bba.dec`